File: sdk/python/src/panproto/_vcs.py
"""VCS (Version Control System) for schema evolution.

Provides a git-like API for versioning schemas with branches,
commits, merges, and blame.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, TypedDict, final

from ._errors import WasmError
from ._msgpack import unpack_from_wasm
from ._wasm import WasmHandle, WasmModule, create_handle

if TYPE_CHECKING:
    from ._schema import BuiltSchema


# ---------------------------------------------------------------------------
# Types
# ---------------------------------------------------------------------------


class VcsLogEntry(TypedDict):
    """A commit log entry."""

    message: str
    author: str
    timestamp: int
    protocol: str


class VcsStatus(TypedDict):
    """Repository status."""

    branch: str | None
    head_commit: str | None


class VcsOpResult(TypedDict):
    """VCS operation result."""

    success: bool
    message: str


class VcsBlameResult(TypedDict):
    """Blame result with commit info."""

    commit_id: str
    author: str
    timestamp: int
    message: str


# ---------------------------------------------------------------------------
# VcsRepository
# ---------------------------------------------------------------------------


@final
class VcsRepository:
    """An in-memory VCS repository for schema evolution.

    Implements the context-manager protocol for automatic cleanup of
    the WASM-side resource.

    Parameters
    ----------
    handle : WasmHandle
        WASM handle for the repository.
    protocol_name : str
        The protocol this repository tracks.
    wasm : WasmModule
        The owning WASM module.

    Examples
    --------
    >>> with VcsRepository.init("atproto", wasm) as repo:
    ...     repo.add(schema)
    ...     repo.commit("initial schema", "alice")
    ...     log = repo.log()
    """

    __slots__ = ("_handle", "_protocol_name", "_wasm")

    def __init__(
        self,
        handle: WasmHandle,
        protocol_name: str,
        wasm: WasmModule,
    ) -> None:
        self._handle: WasmHandle = handle
        self._protocol_name: str = protocol_name
        self._wasm: WasmModule = wasm

    @classmethod
    def init(cls, protocol_name: str, wasm: WasmModule) -> VcsRepository:
        """Initialize a new in-memory repository.

        Parameters
        ----------
        protocol_name : str
            The protocol this repository tracks.
        wasm : WasmModule
            The WASM module.

        Returns
        -------
        VcsRepository
            A new VCS repository.
        """
        raw_handle = wasm.vcs_init(protocol_name.encode())
        handle = create_handle(raw_handle, wasm)
        return cls(handle, protocol_name, wasm)

    @property
    def protocol_name(self) -> str:
        """The protocol name this repository tracks."""
        return self._protocol_name

    @property
    def wasm_handle(self) -> WasmHandle:
        """The underlying WASM handle (internal use only)."""
        return self._handle

    def add(self, schema: BuiltSchema) -> dict[str, str]:
        """Stage a schema for the next commit.

        Parameters
        ----------
        schema : BuiltSchema
            The built schema to stage.

        Returns
        -------
        dict[str, str]
            A dict with ``schema_id`` key.
        """
        try:
            result_bytes = self._wasm.vcs_add(
                self._handle.id,
                schema.wasm_handle.id,
            )
            return unpack_from_wasm(result_bytes)  # type: ignore[return-value]
        except Exception as exc:
            raise WasmError(f"vcs_add failed: {exc}") from exc

    def commit(self, message: str, author: str) -> bytes:
        """Create a commit from the current staging area.

        Parameters
        ----------
        message : str
            The commit message.
        author : str
            The commit author.

        Returns
        -------
        bytes
            Raw commit result bytes.
        """
        try:
            return self._wasm.vcs_commit(
                self._handle.id,
                message.encode(),
                author.encode(),
            )
        except Exception as exc:
            raise WasmError(f"vcs_commit failed: {exc}") from exc

    def log(self, count: int = 50) -> list[VcsLogEntry]:
        """Walk the commit log from HEAD.

        Parameters
        ----------
        count : int
            Maximum number of log entries (default: 50).

        Returns
        -------
        list[VcsLogEntry]
            List of commit log entries.
        """
        try:
            result_bytes = self._wasm.vcs_log(self._handle.id, count)
            return unpack_from_wasm(result_bytes)  # type: ignore[return-value]
        except Exception as exc:
            raise WasmError(f"vcs_log failed: {exc}") from exc

    def status(self) -> VcsStatus:
        """Get the repository status.

        Returns
        -------
        VcsStatus
            Current branch and HEAD commit info.
        """
        try:
            result_bytes = self._wasm.vcs_status(self._handle.id)
            return unpack_from_wasm(result_bytes)  # type: ignore[return-value]
        except Exception as exc:
            raise WasmError(f"vcs_status failed: {exc}") from exc

    def diff(self) -> dict[str, list[dict[str, str]]]:
        """Get diff information for the repository.

        Returns
        -------
        dict
            Diff result with branch info.
        """
        try:
            result_bytes = self._wasm.vcs_diff(self._handle.id)
            return unpack_from_wasm(result_bytes)  # type: ignore[return-value]
        except Exception as exc:
            raise WasmError(f"vcs_diff failed: {exc}") from exc

    def branch(self, name: str) -> VcsOpResult:
        """Create a new branch at the current HEAD.

        Parameters
        ----------
        name : str
            The branch name.

        Returns
        -------
        VcsOpResult
            Operation result.
        """
        try:
            result_bytes = self._wasm.vcs_branch(
                self._handle.id,
                name.encode(),
            )
            return unpack_from_wasm(result_bytes)  # type: ignore[return-value]
        except Exception as exc:
            raise WasmError(f"vcs_branch failed: {exc}") from exc

    def checkout(self, target: str) -> VcsOpResult:
        """Checkout a branch.

        Parameters
        ----------
        target : str
            The branch name to checkout.

        Returns
        -------
        VcsOpResult
            Operation result.
        """
        try:
            result_bytes = self._wasm.vcs_checkout(
                self._handle.id,
                target.encode(),
            )
            return unpack_from_wasm(result_bytes)  # type: ignore[return-value]
        except Exception as exc:
            raise WasmError(f"vcs_checkout failed: {exc}") from exc

    def merge(self, branch_name: str) -> VcsOpResult:
        """Merge a branch into the current branch.

        Parameters
        ----------
        branch_name : str
            The branch to merge.

        Returns
        -------
        VcsOpResult
            Operation result.
        """
        try:
            result_bytes = self._wasm.vcs_merge(
                self._handle.id,
                branch_name.encode(),
            )
            return unpack_from_wasm(result_bytes)  # type: ignore[return-value]
        except Exception as exc:
            raise WasmError(f"vcs_merge failed: {exc}") from exc

    def stash(self) -> VcsOpResult:
        """Stash the current working state.

        Returns
        -------
        VcsOpResult
            Operation result.
        """
        try:
            result_bytes = self._wasm.vcs_stash(self._handle.id)
            return unpack_from_wasm(result_bytes)  # type: ignore[return-value]
        except Exception as exc:
            raise WasmError(f"vcs_stash failed: {exc}") from exc

    def stash_pop(self) -> VcsOpResult:
        """Pop the most recent stash entry.

        Returns
        -------
        VcsOpResult
            Operation result.
        """
        try:
            result_bytes = self._wasm.vcs_stash_pop(self._handle.id)
            return unpack_from_wasm(result_bytes)  # type: ignore[return-value]
        except Exception as exc:
            raise WasmError(f"vcs_stash_pop failed: {exc}") from exc

    def blame(self, vertex_id: str) -> VcsBlameResult:
        """Find which commit introduced a vertex.

        Parameters
        ----------
        vertex_id : str
            The vertex ID to blame.

        Returns
        -------
        VcsBlameResult
            Blame result with commit info.
        """
        try:
            result_bytes = self._wasm.vcs_blame(
                self._handle.id,
                vertex_id.encode(),
            )
            return unpack_from_wasm(result_bytes)  # type: ignore[return-value]
        except Exception as exc:
            raise WasmError(f"vcs_blame failed: {exc}") from exc

    # ------------------------------------------------------------------
    # Context manager / cleanup
    # ------------------------------------------------------------------

    def close(self) -> None:
        """Release the WASM-side repository resource."""
        self._handle.close()

    def __enter__(self) -> VcsRepository:
        return self

    def __exit__(self, *_: object) -> None:
        self.close()

File: sdk/typescript/src/vcs.ts
/**
 * VCS (Version Control System) for schema evolution.
 *
 * Provides a git-like API for versioning schemas with branches,
 * commits, merges, and blame.
 *
 * @module
 */

import type {
  WasmModule,
  VcsLogEntry,
  VcsStatus,
  VcsOpResult,
  VcsBlameResult,
} from './types.js';
import { WasmError } from './types.js';
import { WasmHandle, createHandle } from './wasm.js';
import { unpackFromWasm } from './msgpack.js';
import type { BuiltSchema } from './schema.js';

const encoder = new TextEncoder();

/**
 * An in-memory VCS repository for schema evolution.
 *
 * Implements `Disposable` for automatic cleanup of the WASM-side resource.
 *
 * @example
 * ```typescript
 * using repo = Repository.init('atproto', wasm);
 * repo.add(schema);
 * repo.commit('initial schema', 'alice');
 * const log = repo.log();
 * ```
 */
export class Repository implements Disposable {
  readonly #handle: WasmHandle;
  readonly #wasm: WasmModule;
  readonly #protocolName: string;

  private constructor(handle: WasmHandle, protocolName: string, wasm: WasmModule) {
    this.#handle = handle;
    this.#protocolName = protocolName;
    this.#wasm = wasm;
  }

  /**
   * Initialize a new in-memory repository.
   *
   * @param protocolName - The protocol this repository tracks
   * @param wasm - The WASM module
   * @returns A new disposable Repository
   */
  static init(protocolName: string, wasm: WasmModule): Repository {
    const nameBytes = encoder.encode(protocolName);
    const rawHandle = wasm.exports.vcs_init(nameBytes);
    const handle = createHandle(rawHandle, wasm);
    return new Repository(handle, protocolName, wasm);
  }

  /** The protocol name this repository tracks. */
  get protocolName(): string {
    return this.#protocolName;
  }

  /** The underlying WASM handle. Internal use only. */
  get _handle(): WasmHandle {
    return this.#handle;
  }

  /**
   * Stage a schema for the next commit.
   *
   * @param schema - The built schema to stage
   * @returns An object with the schema's object ID
   */
  add(schema: BuiltSchema): { schemaId: string } {
    try {
      const resultBytes = this.#wasm.exports.vcs_add(
        this.#handle.id,
        schema._handle.id,
      );
      const result = unpackFromWasm<{ schema_id: string }>(resultBytes);
      return { schemaId: result.schema_id };
    } catch (error) {
      throw new WasmError(
        `vcs_add failed: ${error instanceof Error ? error.message : String(error)}`,
        { cause: error },
      );
    }
  }

  /**
   * Create a commit from the current staging area.
   *
   * @param message - The commit message
   * @param author - The commit author
   * @returns The commit result
   */
  commit(message: string, author: string): Uint8Array {
    try {
      return this.#wasm.exports.vcs_commit(
        this.#handle.id,
        encoder.encode(message),
        encoder.encode(author),
      );
    } catch (error) {
      throw new WasmError(
        `vcs_commit failed: ${error instanceof Error ? error.message : String(error)}`,
        { cause: error },
      );
    }
  }

  /**
   * Walk the commit log from HEAD.
   *
   * @param count - Maximum number of log entries to return (default: 50)
   * @returns Array of commit log entries
   */
  log(count: number = 50): VcsLogEntry[] {
    try {
      const resultBytes = this.#wasm.exports.vcs_log(this.#handle.id, count);
      return unpackFromWasm<VcsLogEntry[]>(resultBytes);
    } catch (error) {
      throw new WasmError(
        `vcs_log failed: ${error instanceof Error ? error.message : String(error)}`,
        { cause: error },
      );
    }
  }

  /**
   * Get the repository status.
   *
   * @returns Current branch and HEAD commit info
   */
  status(): VcsStatus {
    try {
      const resultBytes = this.#wasm.exports.vcs_status(this.#handle.id);
      return unpackFromWasm<VcsStatus>(resultBytes);
    } catch (error) {
      throw new WasmError(
        `vcs_status failed: ${error instanceof Error ? error.message : String(error)}`,
        { cause: error },
      );
    }
  }

  /**
   * Get diff information for the repository.
   *
   * @returns Diff result with branch info
   */
  diff(): unknown {
    try {
      const resultBytes = this.#wasm.exports.vcs_diff(this.#handle.id);
      return unpackFromWasm(resultBytes);
    } catch (error) {
      throw new WasmError(
        `vcs_diff failed: ${error instanceof Error ? error.message : String(error)}`,
        { cause: error },
      );
    }
  }

  /**
   * Create a new branch at the current HEAD.
   *
   * @param name - The branch name
   * @returns Operation result
   */
  branch(name: string): VcsOpResult {
    try {
      const resultBytes = this.#wasm.exports.vcs_branch(
        this.#handle.id,
        encoder.encode(name),
      );
      return unpackFromWasm<VcsOpResult>(resultBytes);
    } catch (error) {
      throw new WasmError(
        `vcs_branch failed: ${error instanceof Error ? error.message : String(error)}`,
        { cause: error },
      );
    }
  }

  /**
   * Checkout a branch.
   *
   * @param target - The branch name to checkout
   * @returns Operation result
   */
  checkout(target: string): VcsOpResult {
    try {
      const resultBytes = this.#wasm.exports.vcs_checkout(
        this.#handle.id,
        encoder.encode(target),
      );
      return unpackFromWasm<VcsOpResult>(resultBytes);
    } catch (error) {
      throw new WasmError(
        `vcs_checkout failed: ${error instanceof Error ? error.message : String(error)}`,
        { cause: error },
      );
    }
  }

  /**
   * Merge a branch into the current branch.
   *
   * @param branchName - The branch to merge
   * @returns Operation result
   */
  merge(branchName: string): VcsOpResult {
    try {
      const resultBytes = this.#wasm.exports.vcs_merge(
        this.#handle.id,
        encoder.encode(branchName),
      );
      return unpackFromWasm<VcsOpResult>(resultBytes);
    } catch (error) {
      throw new WasmError(
        `vcs_merge failed: ${error instanceof Error ? error.message : String(error)}`,
        { cause: error },
      );
    }
  }

  /**
   * Stash the current working state.
   *
   * @returns Operation result
   */
  stash(): VcsOpResult {
    try {
      const resultBytes = this.#wasm.exports.vcs_stash(this.#handle.id);
      return unpackFromWasm<VcsOpResult>(resultBytes);
    } catch (error) {
      throw new WasmError(
        `vcs_stash failed: ${error instanceof Error ? error.message : String(error)}`,
        { cause: error },
      );
    }
  }

  /**
   * Pop the most recent stash entry.
   *
   * @returns Operation result
   */
  stashPop(): VcsOpResult {
    try {
      const resultBytes = this.#wasm.exports.vcs_stash_pop(this.#handle.id);
      return unpackFromWasm<VcsOpResult>(resultBytes);
    } catch (error) {
      throw new WasmError(
        `vcs_stash_pop failed: ${error instanceof Error ? error.message : String(error)}`,
        { cause: error },
      );
    }
  }

  /**
   * Find which commit introduced a vertex.
   *
   * @param vertexId - The vertex ID to blame
   * @returns Blame result with commit info
   */
  blame(vertexId: string): VcsBlameResult {
    try {
      const resultBytes = this.#wasm.exports.vcs_blame(
        this.#handle.id,
        encoder.encode(vertexId),
      );
      return unpackFromWasm<VcsBlameResult>(resultBytes);
    } catch (error) {
      throw new WasmError(
        `vcs_blame failed: ${error instanceof Error ? error.message : String(error)}`,
        { cause: error },
      );
    }
  }

  /** Release the WASM-side repository resource. */
  [Symbol.dispose](): void {
    this.#handle[Symbol.dispose]();
  }
}
